core/languages.lua

```diff
@@ -9,30 +13,124 @@ end)
 
 local loadonce = {}
 
+-- BEGIN OMIKHLEIA HACKLANG
+local icu = require("justenoughicu")
+
+-- This small utility could be moved to utilities as SU.forLanguage()...
+-- The input is expected to be a valid BCP47 canonical language.
+-- The idea is to find the "closest" language accepted by the callback:
+-- Loop removing a language specifier until the callback returns a non-nil value
+-- Returns that value and the matched language in that case, or nil no callback matched.
+-- E.g. "xx-Xxxx-XX" will be matched against "xx-Xxxx--XX", "xx-Xxxx", "xx" until one of
+-- these are satisfied.
+-- Returns
+--    nil if not callback could process the language
+--    language, res if a callback could (returning the matched language pattern and the
+--    result of the callback)
+local function forLanguage(langbcp47, callback)
+  while langbcp47 do
+    local res = callback(langbcp47)
+    if res then
+      return res, langbcp47
+    end
+    langbcp47 = langbcp47:match("^(.+)-.*$") -- split at dash (-) and remove last part.
+  end
+  return nil
+end
+-- END OMIKHLEIA HACKLANG
+
 SILE.languageSupport = {
-  languages = {},
+  languages = {}, -- BEGIN OMIKHLEIA HACKLANG
+    -- BAD CODE SMELL WARNING!!!!
+    -- In earlier versions this was used where we now have a "loadonce" table of boolean.
+    -- The change occurred here https://github.com/sile-typesetter/sile/commit/0c5e7f97f3c73ab1b6cd7aee0afca4a59c447cd9
+    -- So this table is not handled as it was, and shouldn't be used! But:
+    --   - font.lua uses it at one point...
+    --   - so does languages/kn.lua
+    --   - and also packages/complex-spaces
+    -- END OMIKHLEIA HACKLANG
   loadLanguage = function (language)
     language = language or SILE.settings:get("document.language")
-    language = cldr.locales[language] and language or "und"
+    -- BEGIN OMIKHLEIA HACKLANG
+    -- Either done too soon or plain wrong (a BCP47 language can e.g. contain a script for instance, sr-Latn
+    -- and I don't see that in https://github.com/alerque/cldr-lua/blob/master/cldr/data/locales.lua
+    -- I don't really know why we would want that CLDR filtering here BTW...
+    -- language = cldr.locales[language] and language or "und"
+
+    -- The user may have set document.language to anything, let's ensure a canonical
+    -- BCP47 language...
+    if language ~= "und" then
+      language = icu.canonicalize_language(language)
```

Comment:
Does this actually parse and match the closes canonical language or just validate the form (see [these comments](https://github.com/sile-typesetter/sile/issues/1368#issuecomment-1327941413))? If the latter I would propose we actually need the former, and the right place to add that functionality would actually be the [cldr-lua](https://github.com/alerque/cldr-lua) library not SILE.

Reply:
It does what is implemented in the SILE wrapper i.e. several things:
- Convert (possibly poor-formatted) BCP47 to correct Locale format.
  - This does _not_ perform any kind of validation. E.g. `ukuou` remains `ukuou` (it's _not_ forbidden by any standard, actually, to design "new language names"); `en-Cyrl` becomes `en_Cyrl` (well, English in Cyrillic script might not really exist, but likewise, why not).
  - It takes care of normalizing the format (capitals, etc.). E.g. `en-us` becomes `en_US`
  - Unrecognized elements are skipped
  - Note that it has actually some provision for scripts, variants, collations, etc. `fr-u-ks-level2-kn` becomes `fr@colnumeric=yes;colstrength=secondary`... I have _not_ taken advantage of this in the collation-enabled sorting, but we could possibly... (there's an interesting question then, but it's a slightly different topic).
- Then _minimizes_ it according to the "canonical" rules for subtags, e.g. (refer to ICU for more details, these are just examples)
  - `en_US` becomes `en` (that's the default implication), while `en_GB` remains itself.
  - Likewise `fr_FR` becomes `fr` (but obviously `fr_CA` etc. stay)
  - `sr_Cyrl` becomes `sr` (again, the default implication is Cyrillic script), while `sr_Latn` stays, etc.
  - Likewise `zh_Hans` = `zh` (the norm, again), and `zh-Hant` stays itself. 
- Convert back to BCP47
  - So we are back with a well formed minimized or "canonical" BCP47

In the end, to keep the same kind of examples:
`en-us` => `en` 
`en-gb` = `en-GB` 
`sr-Cyrl` => `sr`
`sr-Latn` => `sr-Latn`
`fr-u-ks-level2-kn` = `fr-u-kn-true-ks-level2` (yep, it also "normalizes" the optional ICU BCP47 extensions)

All of these are "valid".

What you are asking is something else, the _presumed_ validity of a known language... Well, it dosen't do that.
`ukuo-Kuou` passes (it would be the "Ukuo" language in "Kuou" script. I just invented it)
`en-Cyrl` passes, as stated (it would be the English language in Cyrillic script. Perhaps someone even invented it one day).

But do we have to really care about non-existing combinations?
- They wouldn't have any hyphenation rules anyway
- They wouldn't have any i18n translation keys... Erm, wait, until someone decides to provide them ;-)

Hopefully maybe...
`tlh-Piqd` passes... that's Klingon in the pIqaD script. Yep, it does exist "officially", being even registered in ISO 15924.

This being said, if you want to filter "invalid" languages in **cldr-lua**, you have huge tables to maintain.
- All scripts from ISO 15924 https://www.unicode.org/iso15924/iso15924-codes.html
- Lots of languages, regional variants, etc. https://icu4c-demos.unicode.org/icu-bin/locexp

Currently **cldr-lua** is aligned or CLDR 36 from 2019? We are in 2022, use CLDR 42 at least...

Reply:
The discussion on ICU "canonicalize" also relates to https://github.com/sile-typesetter/sile/issues/276 (when it was implemented).

Reply:
Thanks, that helps a lot!

> But do we have to really care about non-existing combinations?

No. Nor do we care about invalid languages. By the same token we don't care about **valid** languages if we don't have any support for them. What we want is to fall back to the closest thing *we support* in a given context, and that might not be the same for all contexts. A font might support a more specific locale than we have different hyphenation rules or localized strings for.